### PR TITLE
Handle instantiated metrics for BinarySpaceTree.

### LINKED ISSUE #1579: Binary space tree (and subtypes) can't use non static metrics
Reading the documentation it seems I should be able to use a non-static metric with state with neighbor searches and range queries.  However, when I try to setup a tree (specifically a BallTree) with a metric with a non-static Evaluate function I get errors pointing to

https://github.com/mlpack/mlpack/blob/dabd108b38c8f58648c0b137dd60c8e9cf51f243/src/mlpack/core/tree/binary_space_tree/binary_space_tree_impl.hpp#L797-L799

Which of course makes sense, as it's calling a static Evaluate.

Is there a work around or a fix in the works?

### PR DESCRIPTION
**Pull request by rcurtin**
This fixes #1579.

@reverendbedford: sorry, the patch was just a little more complex than I thought.  You can use this branch to fix the issue, or wait until it is merged.  Thanks again for pointing out the issue. :+1:

### CODE REVIEW

File: src/mlpack/core/metrics/mahalanobis_distance_impl.hpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /***
- * @file mahalanobis_distance.cc
+ * @file mahalanobis_distance.cpp
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/core/metrics/mahalanobis_distance_impl.hpp]: Oops, nice catch, thanks!

**Comment by rcurtin** [comment]
Sounds good, I will go ahead and merge this in 3 days to leave time for any other comments.

**Comment by reverendbedford** [comment]
Thanks @rcurtin this solves my problem!

File: src/mlpack/core/metrics/mahalanobis_distance.hpp [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /***
- * @file mahalanobis_dstance.h
+ * @file mahalanobis_dstance.hpp
```
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/core/metrics/mahalanobis_distance.hpp]: Oops, thanks for pointing it out! :)

**Comment by rcurtin** [comment]
@reverendbedford: great to hear this solves your problem!  If you have any more problems in the future please feel free to open an issue and we can try to get it fixed. :)

**Pull request merged**